src/textfont.js

```diff
@@ -226,6 +246,18 @@ export class TextFont {
     return (metric.advanceWidth / this.resolution) * this.pointsToPixels;
   }
 
+  getWidthForString(s) {
+    const key = this.textHashBase + '-' + s;
```

Comment:
The key here can blow up memory. I suggest a two-level hash like `TextFont.widthCache[fontName][text]`.

Reply:
is a [n*m] hash less efficient in some way than 2 level [n][m]?  It seems like the memory usage would be the same.

Reply:
Yes, because you have huge key prefixes which are duplicated for each entry. With 2-level, you don't duplicate the 'font-size-weight' part of the key.

Reply:
An alternative would be to SHA160 hash the key, but I'm not sure that JS has a portable hashing library.

Reply:
You are right!  Updated to use 2 level cache hash.